File: apps/api-dotnet/Auth/AuthService.cs
using System.Security.Cryptography;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using Sarh.Api.Common.Errors;
using Sarh.Api.Data;
using Sarh.Api.Data.Entities;

namespace Sarh.Api.Auth;

public sealed class SignInRequest
{
    public string Email { get; set; } = "";
    public string Password { get; set; } = "";
}

public sealed class SignInWithPinRequest
{
    public string DigitalIdNumber { get; set; } = "";
    public string Pin { get; set; } = "";
}

public sealed class SignInUser
{
    public required string Id { get; init; }
    public required string Email { get; init; }
    public required string Role { get; init; }
    public string? OfficerId { get; init; }
    public string? CitizenId { get; init; }
}

public sealed class SignInResponse
{
    public required string AccessToken { get; init; }
    public required string RefreshToken { get; init; }
    public string TokenType { get; init; } = "bearer";
    public required int ExpiresIn { get; init; }
    public required SignInUser User { get; init; }
}

public sealed class AuthService(SarhDbContext db, JwtTokenService jwt, ILogger<AuthService> log)
{
    public async Task<SignInResponse> SignInWithPinAsync(SignInWithPinRequest dto, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(dto.DigitalIdNumber) || string.IsNullOrWhiteSpace(dto.Pin))
        {
            throw SarhException.Validation(
                "رقم الهوية الرقمية ورمز PIN مطلوبان.",
                "Digital ID number and PIN are required.");
        }

        var did = dto.DigitalIdNumber.Trim();
        var card = await db.DigitalIdCards.AsNoTracking()
            .Where(c => c.DigitalIdNumber == did)
            .FirstOrDefaultAsync(ct);

        // "No such card", "no PIN set" and "wrong PIN" all return the SAME generic
        // message on purpose — distinguishing them would let an attacker enumerate
        // which digital IDs exist (and which lack a PIN) and confirm PINs. On the
        // card-missing / no-PIN paths we still run a bcrypt verify against a
        // throwaway hash so the response time matches the real-PIN path and can't
        // be used as a timing oracle. The precise reason is logged server-side so
        // we can still diagnose a failed login from the API log.
        if (card is null)
        {
            BCrypt.Net.BCrypt.Verify(dto.Pin, DummyPinHash);
            log.LogInformation("PIN login rejected: no card for digital_id_number '{Did}'", did);
            throw InvalidPinCredentials();
        }

        if (string.IsNullOrEmpty(card.PinHash))
        {
            BCrypt.Net.BCrypt.Verify(dto.Pin, DummyPinHash);
            log.LogInformation("PIN login rejected: card '{Did}' has no PIN set (officer must reset PIN)", did);
            throw InvalidPinCredentials();
        }

        if (!BCrypt.Net.BCrypt.Verify(dto.Pin, card.PinHash))
        {
            log.LogInformation("PIN login rejected: wrong PIN for card '{Did}'", did);
            throw InvalidPinCredentials();
        }

        // PIN verified — now safe to reveal card-state details to the legitimate
        // holder. An inactive (frozen/revoked/expired/lost) card can't sign in.
        if (card.Status is not "active")
        {
            log.LogInformation("PIN login rejected: card '{Did}' is '{Status}'", did, card.Status);
            throw CardNotActive(card.Status);
        }

        var citizen = await db.Citizens.AsNoTracking()
            .Where(c => c.Id == card.CitizenId)
            .FirstOrDefaultAsync(ct)
            ?? throw SarhException.Unauthorized();

        var authUser = citizen.AuthUserId is Guid auid
            ? await db.AuthUsers.AsNoTracking().Where(u => u.Id == auid).FirstOrDefaultAsync(ct)
            : null;

        var payload = new SarhJwtPayload
        {
            Sub = (authUser?.Id ?? citizen.Id).ToString(),
            Email = authUser?.Email ?? $"{did}@digital-id.sarh.ly",
            SarhRole = "citizen",
            CitizenId = citizen.Id.ToString(),
            OfficerId = null,
        };
        var (token, expiresIn) = jwt.SignAccessToken(payload);

        var refresh = Convert.ToBase64String(RandomNumberGenerator.GetBytes(48))
            .Replace("+", "-").Replace("/", "_").TrimEnd('=');

        return new SignInResponse
        {
            AccessToken = token,
            RefreshToken = refresh,
            ExpiresIn = expiresIn,
            User = new SignInUser
            {
                Id = (authUser?.Id ?? citizen.Id).ToString(),
                Email = authUser?.Email ?? $"{did}@digital-id.sarh.ly",
                Role = "citizen",
                OfficerId = null,
                CitizenId = citizen.Id.ToString(),
            },
        };
    }

    // A valid bcrypt hash (work factor 10) used only to equalize the timing of
    // the "no card" / "no PIN" rejection paths with a real PIN verification, so
    // the response time can't reveal whether a digital ID exists.
    private static readonly string DummyPinHash =
        BCrypt.Net.BCrypt.HashPassword("sarh-timing-equalizer", 10);

    // Generic invalid-credentials for the PIN flow. Used for "no such card",
    // "no PIN set" and "wrong PIN" so none can be told apart by a caller.
    private static SarhException InvalidPinCredentials() =>
        new(401, "ERR_INVALID_CREDENTIALS",
            "رقم الهوية الرقمية أو رمز PIN غير صحيح.",
            "Digital ID number or PIN is incorrect.");

    // Card-state message shown only after the PIN has been verified (so it never
    // leaks to someone who doesn't already hold the card).
    private static SarhException CardNotActive(string status)
    {
        var ar = status switch
        {
            "frozen" => "بطاقتك مجمّدة مؤقتاً. يرجى مراجعة مكتب الإصدار.",
            "revoked" => "بطاقتك ملغاة ولا يمكن استخدامها. يرجى مراجعة مكتب الإصدار.",
            "expired" => "انتهت صلاحية بطاقتك. يرجى مراجعة مكتب الإصدار لإعادة إصدارها.",
            "lost" => "بطاقتك مُبلّغ عنها كمفقودة. يرجى مراجعة مكتب الإصدار.",
            _ => "بطاقتك غير مفعّلة حالياً. يرجى مراجعة مكتب الإصدار.",
        };
        return new SarhException(403, "ERR_CARD_NOT_ACTIVE", ar,
            $"Your card is not active (status: {status}).");
    }

    public async Task<SignInResponse> SignInAsync(SignInRequest dto, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(dto.Email) || string.IsNullOrWhiteSpace(dto.Password))
        {
            throw SarhException.Validation(
                "البريد الإلكتروني وكلمة المرور مطلوبان.",
                "Email and password are required.");
        }

        var email = dto.Email.Trim().ToLowerInvariant();
        var user = await db.AuthUsers.AsNoTracking()
            .Where(u => u.Email == email)
            .FirstOrDefaultAsync(ct)
            ?? throw SarhException.Unauthorized();

        var ok = BCrypt.Net.BCrypt.Verify(dto.Password, user.EncryptedPassword);
        if (!ok) throw SarhException.Unauthorized();

        var officer = await db.Officers.AsNoTracking()
            .Where(o => o.AuthUserId == user.Id)
            .FirstOrDefaultAsync(ct);

        var (role, citizenId) = ResolveRoleAndCitizen(user, officer);
        if (role is null) throw SarhException.Unauthorized();

        var payload = new SarhJwtPayload
        {
            Sub = user.Id.ToString(),
            Email = user.Email,
            SarhRole = role,
            CitizenId = citizenId,
            OfficerId = officer?.IsActive == true ? officer.Id.ToString() : null,
            RegionId = officer?.RegionId,
            MunicipalityId = officer?.MunicipalityId,
        };
        var (token, expiresIn) = jwt.SignAccessToken(payload);

        // Stamp last_sign_in_at (best-effort; raw SQL avoids tracking churn).
        try
        {
            await db.Database.ExecuteSqlRawAsync(
                "UPDATE auth_users SET last_sign_in_at = SYSDATETIMEOFFSET() WHERE id = {0}",
                new object[] { user.Id }, ct);
        }
        catch { /* non-critical */ }

        var refresh = Convert.ToBase64String(RandomNumberGenerator.GetBytes(48))
            .Replace("+", "-").Replace("/", "_").TrimEnd('=');

        return new SignInResponse
        {
            AccessToken = token,
            RefreshToken = refresh,
            ExpiresIn = expiresIn,
            User = new SignInUser
            {
                Id = user.Id.ToString(),
                Email = user.Email,
                Role = role,
                OfficerId = officer?.IsActive == true ? officer.Id.ToString() : null,
                CitizenId = citizenId,
            },
        };
    }

    private static (string? Role, string? CitizenId) ResolveRoleAndCitizen(AuthUser user, Officer? officer)
    {
        string? appRole = null;
        string? citizenId = null;
        if (!string.IsNullOrWhiteSpace(user.RawAppMetaData))
        {
            try
            {
                using var doc = JsonDocument.Parse(user.RawAppMetaData);
                if (doc.RootElement.TryGetProperty("sarh_role", out var r) && r.ValueKind == JsonValueKind.String)
                    appRole = r.GetString();
                if (doc.RootElement.TryGetProperty("citizen_id", out var c) && c.ValueKind == JsonValueKind.String)
                    citizenId = c.GetString();
            }
            catch { /* ignore malformed JSON */ }
        }
        var role = (officer?.IsActive == true ? officer.Role : null) ?? appRole;
        return (role, citizenId);
    }
}

File: apps/api/src/auth/auth.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { randomUUID, randomBytes } from 'node:crypto';
import * as bcrypt from 'bcrypt';
import { SupabaseService } from '../supabase/supabase.service';
import { SijilliErrors } from '../common/errors/error-envelope';
import { SignInDto } from './dto/sign-in.dto';
import { InviteOfficerDto } from './dto/invite-officer.dto';
import { JwtService, SijilliJwtPayload } from './jwt.service';

export interface SignInResult {
  access_token: string;
  refresh_token: string;
  token_type: 'bearer';
  expires_in: number;
  user: {
    id: string;
    email: string | null;
    role: string | null;
    officer_id: string | null;
    citizen_id: string | null;
  };
}

export interface InvitedOfficer {
  id: string;
  auth_user_id: string;
  email: string;
  employee_no: string;
  role: string;
  full_name_ar: string;
  recovery_token: string; // one-shot password reset token
}

interface AuthUserRow {
  id: string;
  email: string;
  encrypted_password: string;
  email_confirmed_at: Date | null;
  raw_app_meta_data: Record<string, unknown>;
  raw_user_meta_data: Record<string, unknown>;
}

const BCRYPT_COST = 10;

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    private readonly supabase: SupabaseService,
    private readonly jwt: JwtService,
  ) {}

  // ---------------- Sign in ------------------------------------------
  async signIn(dto: SignInDto): Promise<SignInResult> {
    const lookup = await this.supabase.admin
      .from<AuthUserRow>('auth_users')
      .select('id, email, encrypted_password, email_confirmed_at, raw_app_meta_data, raw_user_meta_data')
      .eq('email', dto.email.toLowerCase())
      .maybeSingle();
    if (lookup.error) throw SijilliErrors.upstream(lookup.error.message);
    if (!lookup.data) throw SijilliErrors.unauthorized();

    const user = lookup.data;
    const ok = await bcrypt.compare(dto.password, user.encrypted_password);
    if (!ok) throw SijilliErrors.unauthorized();

    // Resolve officer / citizen ids for the response and the JWT.
    const officer = await this.supabase.admin
      .from<{ id: string; role: string; region_id: number | null; municipality_id: number | null; permissions: Record<string, unknown> | null; is_active: boolean }>('officers')
      .select('id, role, region_id, municipality_id, permissions, is_active')
      .eq('auth_user_id', user.id)
      .maybeSingle();

    const appMeta = user.raw_app_meta_data ?? {};
    const role =
      (officer.data?.is_active ? officer.data.role : null) ??
      (appMeta['sijilli_role'] as string | undefined) ??
      null;
    const citizenId = (appMeta['citizen_id'] as string | undefined) ?? null;

    if (!role) throw SijilliErrors.unauthorized();

    const payload: SijilliJwtPayload = {
      sub: user.id,
      email: user.email,
      sijilli_role: role,
      citizen_id: citizenId ?? undefined,
      officer_id: officer.data?.id,
      region_id: officer.data?.region_id ?? null,
      municipality_id: officer.data?.municipality_id ?? null,
      permissions: (officer.data?.permissions as Record<string, boolean | string | number> | null) ?? null,
    };

    const { token, expiresIn } = this.jwt.signAccessToken(payload);
    const refresh = randomBytes(48).toString('base64url');

    // Stamp last_sign_in_at (best-effort, ignore errors).
    await this.supabase.admin
      .from('auth_users')
      .update({ last_sign_in_at: new Date().toISOString() })
      .eq('id', user.id);

    return {
      access_token: token,
      refresh_token: refresh,
      token_type: 'bearer',
      expires_in: expiresIn,
      user: {
        id: user.id,
        email: user.email,
        role,
        officer_id: officer.data?.id ?? null,
        citizen_id: citizenId,
      },
    };
  }

  // ---------------- Invite officer -----------------------------------
  async inviteOfficer(dto: InviteOfficerDto): Promise<InvitedOfficer> {
    const authUserId = randomUUID();
    // Random initial password — caller emails the recovery_token.
    const tempPwd = randomBytes(24).toString('base64url');
    const hash = await bcrypt.hash(tempPwd, BCRYPT_COST);

    const created = await this.supabase.admin
      .from<AuthUserRow>('auth_users')
      .insert({
        id: authUserId,
        email: dto.email.toLowerCase(),
        encrypted_password: hash,
        email_confirmed_at: new Date().toISOString(),
        raw_app_meta_data: {
          sijilli_role: dto.role,
          permissions: dto.permissions ?? {},
        },
        raw_user_meta_data: {
          full_name_ar: dto.full_name_ar,
        },
      })
      .select('id, email')
      .single();
    if (created.error || !created.data) {
      if (created.error?.code === '23505') {
        throw SijilliErrors.conflict(
          'البريد الإلكتروني مستخدم مسبقاً.',
          'Email already exists.',
        );
      }
      throw SijilliErrors.upstream(`Failed to create auth user: ${created.error?.message ?? 'unknown'}`);
    }

    const inserted = await this.supabase.admin
      .from<{ id: string; auth_user_id: string; email: string; employee_no: string; role: string; full_name_ar: string }>('officers')
      .insert({
        auth_user_id: authUserId,
        employee_no: dto.employee_no,
        full_name_ar: dto.full_name_ar,
        full_name_en: dto.full_name_en ?? null,
        role: dto.role,
        region_id: dto.region_id ?? null,
        municipality_id: dto.municipality_id ?? null,
        phone: dto.phone ?? null,
        email: dto.email.toLowerCase(),
        permissions: dto.permissions ?? {},
        is_active: true,
      })
      .select('id, auth_user_id, email, employee_no, role, full_name_ar')
      .single();

    if (inserted.error || !inserted.data) {
      // Roll back the auth user so we don't orphan it.
      await this.supabase.admin.from('auth_users').delete().eq('id', authUserId);
      if (inserted.error?.code === '23505') {
        throw SijilliErrors.conflict(
          'الرقم الوظيفي أو البريد الإلكتروني مستخدم مسبقاً.',
          'Employee number or email already exists.',
        );
      }
      throw SijilliErrors.upstream(`Failed to insert officer: ${inserted.error?.message ?? 'unknown'}`);
    }

    return {
      id: inserted.data.id,
      auth_user_id: inserted.data.auth_user_id,
      email: inserted.data.email,
      employee_no: inserted.data.employee_no,
      role: inserted.data.role,
      full_name_ar: inserted.data.full_name_ar,
      recovery_token: tempPwd,
    };
  }
}
